webapp/channels/src/components/tooltip/index.tsx

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {
+    useFloating,
+    autoUpdate,
+    offset,
+    useHover,
+    useFocus,
+    useDismiss,
+    useRole,
+    useInteractions,
+    arrow,
+    FloatingPortal,
+    autoPlacement,
+    useTransitionStyles,
+    FloatingArrow,
+} from '@floating-ui/react';
+import type {Placement} from '@floating-ui/react';
+import React, {cloneElement, isValidElement, useMemo, useRef, useState, memo} from 'react';
+import type {ReactNode} from 'react';
+import {defineMessage} from 'react-intl';
+
+import TooltipContent from 'components/tooltip/tooltip_content';
+import type {ShortcutDefinition} from 'components/tooltip/tooltip_shortcut';
+
+import {Constants} from 'utils/constants';
+
+import './tooltip.scss';
+
+const ARROW_WIDTH = 10;
+const ARROW_HEIGHT = 6;
+const ARROW_OFFSET = 8;
+
+const TOOLTIP_REST_TIME_BEFORE_OPEN = 100;
+const TOOLTIP_APPEAR_DURATION = 250;
+const TOOLTIP_DISAPPEAR_DURATION = 200;
+
+const DEFAULT_PLACEMENT: Placement = 'top';
+const DEFAULT_ALLOWED_PLACEMENTS: Placement[] = ['top', 'bottom'];
+
+export const ShortcutKeys = {
+    alt: defineMessage({
+        id: 'shortcuts.generic.alt',
+        defaultMessage: 'Alt',
+    }),
+    cmd: '⌘',
+    ctrl: defineMessage({
+        id: 'shortcuts.generic.ctrl',
+        defaultMessage: 'Ctrl',
+    }),
+    option: '⌥',
+    shift: defineMessage({
+        id: 'shortcuts.generic.shift',
+        defaultMessage: 'Shift',
+    }),
+};
+
+interface Options {
+
+    /**
+     * The placement of the tooltip. Defaults placement is top.
+     */
+    placement?: Placement;
+
+    /**
+     * Whether to automatically determine the placement of the tooltip.
+     * ref: https://floating-ui.com/docs/autoPlacement
+     */
+    autoPlacement?: boolean;
+
+    /**
+     * The allowed placements of the tooltip. This should be used in conjunction with `autoPlacement`.
+     */
+    allowedPlacements?: Placement[];
+
+    /**
+     * Callback fired when the tooltip is opened or closed.
+     */
+    onChange?: (open: boolean) => void;
+}
+
+interface Props {
+    children: ReactNode;
```

Comment:
If this type is `ReactElement`, it should avoid the need to check this with `isValidElement`. I think `ReactElement` type prevents `children` from being null, a string, or an array

Reply:
changing ReactElement works but as per react docs `type PropsWithChildren<P> = P & { children?: ReactNode | undefined };` also in the latest comment we arent checking if its empty.

Reply:
I don't know if that `PropsWithChildren` is actually a type we want to use for things where this component cares specifically what it's children are. Since we can restrict this component to take only a single child with the type system, let's do that here